### DISCUSSION TITLE
Does not work with bazel-remote: bad request

### DISCUSSION THREAD

**Discussion by liaoqinghai** [ccache/ccache#1516] [Support]
### How to reproduce ###

1.  setup `bazel-remote-2.4.4-linux-x86_64` as wiki
2.  setup ccache config as wiki:  setup  remote_storage, layout=bazel, remote_only=true (for test)
3.  open wireshark to capture network data (for debug)
4. compile some thing

### Actual behavior ###

remote cache function not work

Has a `demo.cpp` : 
```cpp
#include <iostream>

int main(int argc, char** argv)
{
  std::cout << "hello world!" << std::endl;
  return 0;
}
```
called with:
```shell
# alreay set CCACHE_DIR
$ ccache g++ -c demo.cpp
```
use wireshark to capture the data, find `Bad Request` , as the following image:
![image](https://github.com/user-attachments/assets/10c5f696-2947-43bb-b516-68f7896a35d5)
![image](https://github.com/user-attachments/assets/abcd688a-67d2-47b6-8736-b64cfc5a876e)

### Expected behavior ###

Just work normally

I tried nginx as wiki which  work well.

### Environment ###

ccache version 4.10.2
bazel-remote-2.4.4-linux-x86_64

Ubuntu 20.04
gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0

**Comment by jrosdahl** [comment]
The Wireshark capture indicates that "layout=bazel" isn't in effect, so I think that you didn't manage to configure "layout=bazel" correctly.

How did you configure ccache? An easy way to check what happens is to tell ccache to [produce a log file](https://ccache.dev/manual/4.10.2.html#config_log_file).

**Reply by liaoqinghai** [reply]
my ccache.conf:
```conf
max_size = 10G
compression = true
compression_level = 6
hard_link = false
remote_storage = http://127.0.0.1:8000
layout=bazel
remote_only=true
debug=true
debug_dir=/home/nick/ccache_bazel_demo/cache/debug
```
The bazel-remote terminal output:
```shell
$ ./bazel-remote-2.4.4-linux-x86_64 --config_file ./bazel_remote.cfg
2024/09/23 06:58:49 bazel-remote built with go1.22.0 X:nocoverageredesign from git commit 54d1782d72b291937988edad32c9752abe269d8e.
2024/09/23 06:58:49 Initial RLIMIT_NOFILE cur: 1048576 max: 1048576
2024/09/23 06:58:49 Setting RLIMIT_NOFILE cur: 1048576 max: 1048576
2024/09/23 06:58:49 Storage mode: zstd
2024/09/23 06:58:49 Zstandard implementation: go
2024/09/23 06:58:49 Limiting concurrent file removals to 5000
2024/09/23 06:58:49 Loading existing files in /home/nick/ccache_bazel_demo/bazel_remote.
2024/09/23 06:58:49 Scanning cache directory with 16 goroutines
2024/09/23 06:58:49 Sorting cache files by atime.
2024/09/23 06:58:49 Building LRU index.
2024/09/23 06:58:49 Finished loading disk cache files.
2024/09/23 06:58:49 Authentication: disabled
2024/09/23 06:58:49 Mangling non-empty instance names with AC keys: disabled
2024/09/23 06:58:49 gRPC AC dependency checks: enabled
2024/09/23 06:58:49 experimental gRPC remote asset API: disabled
2024/09/23 06:58:49 Starting gRPC server on address 
2024/09/23 06:58:49 Loaded 0 existing disk cache items.
2024/09/23 06:58:49 Endpoint metrics: disabled
2024/09/23 06:58:49 HTTP AC validation: disabled
2024/09/23 06:58:49 Starting HTTP server on address 0.0.0.0:8000
2024/09/23 06:59:19  GET 400       127.0.0.1 /ac/079pdvb8vjtv4na37apgaefd872shau
2024/09/23 06:59:19  GET 400       127.0.0.1 /c6/cdkasfpoj9c8rarc82sqdoh1i11e5cs
2024/09/23 06:59:19  PUT 400       127.0.0.1 /c6/cdkasfpoj9c8rarc82sqdoh1i11e5cs
^C2024/09/23 07:00:45 Received signal: interrupt, attempting graceful shutdown
2024/09/23 07:00:45 Stopping HTTP server
2024/09/23 07:00:45 Stopping gRPC server
2024/09/23 07:00:45 HTTP server stopped
2024/09/23 07:00:45 gRPC server stopped
```

Here is the demo including output log: [ccache_bazel_demo.zip](https://github.com/user-attachments/files/17093725/ccache_bazel_demo.zip) , I setup the bazel-remote in localhost and my home directory is `/home/nick` .

The same configuration works with nginx setup.

**Reply by jrosdahl** [reply]
Thanks. This is the problem:

> ```
> remote_storage = http://127.0.0.1:8000
> layout=bazel
> ```

It should be `remote_storage = http://127.0.0.1:8000|layout=bazel`. There is no global `layout` configuration option. See also step 2 under _bazel-remote_ on <https://github.com/ccache/ccache/wiki/HTTP-storage#bazel-remote>.

**Reply by liaoqinghai** [reply]
it works.  
Thanks a lot!
